compiler/resolution/generics.cpp

```diff
@@ -179,24 +196,42 @@ checkInfiniteWhereInstantiation(FnSymbol* fn) {
 //
 static void
 checkInstantiationLimit(FnSymbol* fn) {
-  static Map<FnSymbol*,int> instantiationLimitMap;
-
-  // Don't count instantiations on internal modules
-  // nor ones explicitly marked NO_INSTANTIATION_LIMIT.
-  if (fn->getModule() &&
-      fn->getModule()->modTag != MOD_INTERNAL &&
-      !fn->hasFlag(FLAG_NO_INSTANTIATION_LIMIT)) {
+  if (trackInstantiationsForFn(fn)) {
     if (instantiationLimitMap.get(fn) >= instantiation_limit) {
       USR_FATAL_CONT(fn, "Function '%s' has been instantiated too many times",
                      fn->name);
       USR_PRINT("  If this is intentional, try increasing"
                 " the instantiation limit from %d", instantiation_limit);
       USR_STOP();
     }
+    // printf("Incrementing instantiation count for %s (%p)\n", fn->name, fn);
     instantiationLimitMap.put(fn, instantiationLimitMap.get(fn)+1);
   }
 }
 
+void popInstantiationLimit(FnSymbol* fn) {
+  //
+  // varargs functions are not added to the instantiationLimitMap
+  //
+  if (fn->hasFlag(FLAG_EXPANDED_VARARGS)) {
+    return;
+  }
+
+  //
+  // Go from a concrete instantiation to the generic it was based upon (if any)
+  //
+  fn = fn->instantiatedFrom;
```

Comment:
I think it's possible for a function to be instantiated more than once, so I think this part should be in a loop.

Reply:
What do you think about walking up to the "top" instantiatedFrom pointer when inserting functions into this table as well?  I was thinking that we would only put such functions into the table, but it appears that generic varargs functions may result in a few levels of instantiatedFrom (at least, `stringify()` seems to fail a `fn->instantiatedFrom==NULL` assertion in the routine that inserts functions into the table, which may also relate to why I had to special-case varargs functions in the lines just above the ones you've commented on here).

Reply:
Yes I think the instantiation limit should track the "top" generic function